File: app/Sources/LocalCaptionKit/LocalAgreement.swift
import Foundation

/// LocalAgreement-2 interim stabilization (SPEC.md §8.2, glossary): commit only the tokens
/// agreed by the two most recent hypotheses, keeping the rest provisional, so the committed
/// prefix of the live (dimmed) line never flickers backward. Reset per utterance.
public struct LocalAgreement {
    private var previous: [String] = []
    public private(set) var committed: [String] = []

    public init() {}

    public mutating func reset() { previous = []; committed = [] }

    /// Feed the latest hypothesis (word list). The committed prefix grows to the common
    /// prefix of the last two hypotheses and is monotonic (never shrinks).
    /// Returns the committed prefix and the provisional tail.
    @discardableResult
    public mutating func update(_ hypothesis: [String]) -> (committed: [String], provisional: [String]) {
        let agreed = Self.commonPrefix(previous, hypothesis)
        if agreed.count > committed.count { committed = agreed }
        previous = hypothesis

        // Provisional tail = whatever the current hypothesis has beyond the committed prefix.
        let tail = hypothesis.count > committed.count
            ? Array(hypothesis[committed.count...])
            : []
        return (committed, tail)
    }

    static func commonPrefix(_ a: [String], _ b: [String]) -> [String] {
        var out: [String] = []
        for (x, y) in zip(a, b) {
            if x == y { out.append(x) } else { break }
        }
        return out
    }
}

File: windows/src/LocalCaption.Core/Captions/LocalAgreement.cs
namespace LocalCaption.Core.Captions;

/// <summary>
/// LocalAgreement-2 interim stabilisation (SPEC.md §8.2): commit only the tokens agreed by
/// the two most recent hypotheses and keep the rest provisional, so the committed prefix of
/// the live (dimmed) line never flickers backward. Reset per utterance.
/// </summary>
/// <remarks>
/// Requires a <b>fixed</b> audio origin. Moving windows use <see cref="RollingCaption"/>.
/// </remarks>
public sealed class LocalAgreement
{
    private List<string> _previous = [];
    private List<string> _committed = [];

    public IReadOnlyList<string> Committed => _committed;

    public void Reset()
    {
        _previous = [];
        _committed = [];
    }

    /// <summary>
    /// Feed the latest hypothesis. The committed prefix grows to the common prefix of the
    /// last two hypotheses and is monotonic — it never shrinks, even if the newest
    /// hypothesis is shorter or diverges.
    /// </summary>
    public (IReadOnlyList<string> Committed, IReadOnlyList<string> Provisional) Update(
        IReadOnlyList<string> hypothesis)
    {
        var agreed = CommonPrefix(_previous, hypothesis);
        if (agreed.Count > _committed.Count) _committed = agreed;
        _previous = [.. hypothesis];

        // The provisional tail is measured from the COMMITTED length, not from the length
        // of the previous hypothesis.
        var tail = hypothesis.Count > _committed.Count
            ? hypothesis.Skip(_committed.Count).ToList()
            : [];
        return (_committed, tail);
    }

    internal static List<string> CommonPrefix(IReadOnlyList<string> a, IReadOnlyList<string> b)
    {
        var out_ = new List<string>();
        var n = Math.Min(a.Count, b.Count);
        for (var i = 0; i < n && a[i] == b[i]; i++) out_.Add(a[i]);
        return out_;
    }
}
